src/common/range.rs

```diff
@@ -141,278 +147,451 @@ impl ByteRangeSpec {
-            Range::Unregistered(ref unit, ref range_str) => {
-                write!(f, "{}={}", unit, range_str)
-            },
+            }
+            Range::Unregistered { unit, set } => write!(f, "{}={}", unit, set),
         }
     }
 }
 
 impl FromStr for Range {
-    type Err = ::Error;
+    type Err = InvalidRange;
 
-    fn from_str(s: &str) -> ::Result<Range> {
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
         let mut iter = s.splitn(2, '=');
 
         match (iter.next(), iter.next()) {
-            (Some("bytes"), Some(ranges)) => {
-                let ranges = from_comma_delimited(ranges);
-                if ranges.is_empty() {
-                    return Err(::Error::Header);
+            (Some("bytes"), Some(set)) => {
+                let specs: Result<Vec<ByteRangeSpec>, Self::Err> =
+                    set.split(',').map(|spec| spec.parse()).collect();
+                match specs {
+                    Ok(specs) if !specs.is_empty() => Ok(Self::Bytes(specs)),
+                    _ => Err(err()),
                 }
-                Ok(Range::Bytes(ranges))
             }
-            (Some(unit), Some(range_str)) if unit != "" && range_str != "" => {
-                Ok(Range::Unregistered(unit.to_owned(), range_str.to_owned()))
+            (Some(unit), Some(set)) if !unit.is_empty() && !set.is_empty() => {
+                Ok(Self::Unregistered {
+                    unit: unit.to_string(),
+                    set: set.to_string(),
+                })
+            }
+            _ => Err(err()),
+        }
+    }
+}
 
-            },
-            _ => Err(::Error::Header)
+impl std::fmt::Display for ByteRangeSpec {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match *self {
+            ByteRangeSpec::FromTo(first, last) => write!(f, "{}-{}", first, last),
+            ByteRangeSpec::Last(len) => write!(f, "-{}", len),
+            ByteRangeSpec::AllFrom(first) => write!(f, "{}-", first),
         }
     }
 }
 
 impl FromStr for ByteRangeSpec {
-    type Err = ::Error;
-
-    fn from_str(s: &str) -> ::Result<ByteRangeSpec> {
-        let mut parts = s.splitn(2, '-');
-
-        match (parts.next(), parts.next()) {
-            (Some(""), Some(end)) => {
-                end.parse().or(Err(::Error::Header)).map(ByteRangeSpec::Last)
-            },
-            (Some(start), Some("")) => {
-                start.parse().or(Err(::Error::Header)).map(ByteRangeSpec::AllFrom)
-            },
-            (Some(start), Some(end)) => {
-                match (start.parse(), end.parse()) {
-                    (Ok(start), Ok(end)) if start <= end => Ok(ByteRangeSpec::FromTo(start, end)),
-                    _ => Err(::Error::Header)
-                }
+    type Err = InvalidRange;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut iter = s.trim().splitn(2, '-');
+        Ok(
+            match (parse_bound(iter.next())?, parse_bound(iter.next())?) {
```

Comment:
I don't think it makes sense to pass an `Option` to `parse_bound` here. Here are situations where `iter.next()` can return `None` here:
- for the first call, it's impossible. Even for empty string, `splitn` will give at least one element. [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=75e9f7bbca16a22d98d8115b778c4c26), also splitn [docs](https://doc.rust-lang.org/std/primitive.str.html#method.splitn) show this
- for the second call to `iter.next()`, you'll only get `None` if the string doesn't contain `-`. Even in the `foo-` case, you'll get empty string as the second component. [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7b95668da3e90b558e6def97b696eba6)

If the string doesn't contain a `-`, that's an error.

Reply:
oki, I've changed this function back to the previous one which didn't have these issues